<!DOCTYPE html>
<html>
<head>
    <title>UTR Lengthening (UTRL) Analysis parameters</title>
    <meta charset="UTF-8">
    <meta name="author" content="Ãngeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>UTR Lengthening (UTRL) Analysis parameters</b></span></div>
    <p>To evaluate changes in UTR Lengthening (UTRL), you must provide the following information:
    </p>
    <ol class="dlgSection">
        <li>Lengthening range
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Minimum value for differentiation - </b>number of base pairs to call a different UTR length between isoforms.
		All isoforms with UTR lengths differing less than x base pairs (where x is the provided threshold) are considered to have the same UTR length. Once UTR lengths for analysis have been selected, 
		each length is weighted by the relative expression of the isoform (isoform expression / total gene expression), and sumed to provide a weighted 3' or 5' UTR length (wUTR).
            </div>
        </li>
        <li>Analysis Method section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Method - </b>select which analysis method to use to test UTRL:
                <ul class="dlgFields">
                    <li>Wilcoxon - Available for two-group comparisons. Performs a Wilcoxon signed-rank test on the wUTR values between conditions, and provides a p-value indicating whether
			the change in UTR length is globally significant.<br/><br/></li>
		    <li>maSigPro - Available for time-course designs. Performs a gene-level analysis of wUTR values across time-points, and provides a p-value for each gene indicating whether
			the change in UTR length (weighted by each isoform's proportion of expression) is globally significant across time, for each of the series provided (in the case of multiple designs).<br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Analysis Parameters section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Significance Level - </b>specify the significance level required to call a gene as significantly 
											changing UTR length (for time-course desings) or to consider the global UTR length change between conditions significant (for two-group comparisons)
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Polynomial Degree - </b>(time-course only) degree for the maSigPro regression model. The maximum allowed degree is time_points-1.
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Filtering - </b>specify whether or not to filter minor isoforms prior to testing for UTR Lengthening. 
												Recommended to prevent considering UTR lengths from minor isoforms (i.e. very lowly expressed). 
												Two methods are available:
                <ul class="dlgFields">
                    <li>Filtering by proportion - <span style="font-weight: normal;">does not consider isoforms that represent a proportion of gene expression that is below the specified threshold (default: 0.2)</span><br/></li>
                    <li>Filtering by fold expression difference - <span style="font-weight: normal;">compares all isoforms to the major (i.e. the one with highest mean expression across conditions) and filters 
													those with a fold difference above the specified threshold (default: 2)</span><br/><br/></li>
                </ul>
	    </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Max K Clusters - </b>(time-course only) to show the trajectories in plots, maSigPro visualization applies clustering methods to group 
											genes with similar weighted UTR expression trends, and summarizes this in the graphical display. 
											This parameter determines the maximum number of clusters to classify genes into, using wUTR values.
											Enable the "Use mclust for optimal number" option to use a method to dynamically determine the number of clusters.
            </div>
        </li>
    </ol>
    <!-- AddHelp:dfi,edger,dexseq,gui,tips -->
    <br/><br/>
</body>
</html>

