<!DOCTYPE html>

<!--
    @author P. Salguero - psalguero@cipf.es
-->

<html>
    <head>
        <!-- <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Montserrat">  -->
        <title>Start Page</title>
        <meta charset="UTF-8">
	<meta name="author" content="P. Salguero">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <style>            
            html, body {
                border-spacing:0px;
                padding:0px;
                margin:0px;
                font-size: medium;
                font-family: 'Verdana', sans-serif;
                background-color:white;
            }
            .projectText {
                font-size:medium;
                font-weight:normal;
                color: slategray;
            }
            .projectText:hover {
                cursor:pointer;
                color:#a2ca72;
            }
            .cmdText {
                font-weight:bold;
                font-size:large;
                color:slategray;
            }
            .cmdTextNA {
                font-weight:bold;
                font-size:large;
                color:slategray;
            }
            .cmdText:hover {
                cursor:pointer;
                color:#a2ca72;
            }
        </style>
    </head>

    <body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">

        <div style="background-color:white; padding:10px;">
            <span style="font-size:small; color:black;"><b>To cite us visit: </b></br>
            <div style="padding:15px;">
            https://app.tappas.org/downloads
            </div>
            </span>
            <span style="font-size:small; color:black;"><b>Or use the following text: </b></br>
                <div style="padding:15px;">
                tappAS: a comprehensive computational framework for the analysis of the functional impact of differential splicing. </br></br>
                Lorena de la Fuente, Ángeles Arzalluz-Luque, Manuel Tardáguila, Héctor del Risco, Cristina Martí, Sonia Tarazona, Pedro Salguero, Raymond Scott, Ana Alastrue-Agudo, Pablo Bonilla, Jeremy Newman, Lauren McIntyre, Victoria Moreno-Manzano, Ana Cones.</br></br>
                Genome Biology; doi: https://doi.org/10.1186/s13059-020-02028-w
            </div>
            </span>
        </div>
        
        <div style="background-color:white; padding:10px;">
            <span style="font-size:small; color:black;"><b>R Packages: </b></br>
            <div style="padding:15px;">
                    <b>maSigPro:</b>
                    Conesa A, Nueda MJ (2018). maSigPro: Significant Gene Expression Profile Differences in Time Course Gene Expression Data. R package version 1.52.0, http://bioinfo.cipf.es.
                    </br></br>
                    <b>NoiSeq:</b>
                    Tarazona S, Garcia-Alcalde F, Dopazo J, Ferrer A and Conesa A (2011). “Differential expression in RNA-seq: a matter of depth.” Genome Research, 21(12), pp. 4436.
                    </br></br>
                    Tarazona S, Furio-Tari P, Turra D, Pietro AD, Nueda MJ, Ferrer A and Conesa A (2015). “Data quality aware analysis of differential expression in RNA-seq with NOISeq R/Bioc package.” Nucleic Acids Research, 43(21), pp. e140.
                    </br></br>
                    <b>edgerR:</b>
                    Robinson MD, McCarthy DJ and Smyth GK (2010). “edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.” Bioinformatics, 26(1), pp. 139-140.
                    </br></br>
                    McCarthy, J. D, Chen, Yunshun, Smyth and K. G (2012). “Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation.” Nucleic Acids Research, 40(10), pp. 4288-4297.
                    </br></br>
                    <b>DEXSeq:</b>
                    Anders S, Reyes A and Huber W (2012). “Detecting differential usage of exons from RNA-seq data.” Genome Research, 22, pp. 4025. doi: 10.1101/gr.133744.111.
                    </br></br>
                    Reyes A, Anders S, Weatheritt R, Gibson T, Steinmetz L and Huber W (2013). “Drift and conservation of differential exon usage across tissues in primate species.” PNAS, 110, pp. -5. doi: 10.1073/pnas.1307202110.
                    </br></br>
                    <b>Goseq:</b>
                    Young MD, Wakefield MJ, Smyth GK and Oshlack A (2010). “Gene ontology analysis for RNA-seq: accounting for selection bias.” Genome Biology, 11, pp. R14.
                    </br></br>
                    <b>MASS (Cran):</b>
                    Venables, W. N. & Ripley, B. D. (2002) Modern Applied Statistics with S. Fourth Edition. Springer, New York. ISBN 0-387-95457-0
                    </br></br>
                    <b>Ggplot2 (Cran):</b>
                    H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2009.
                    </br></br>
                    <b>Plyr (Cran):</b>
                    Hadley Wickham (2011). The Split-Apply-Combine Strategy for Data Analysis. Journal of Statistical Software, 40(1), 1-29. URL http://www.jstatsoft.org/v40/i01/.
                    </br></br>
                    <b>VennDiagram (Cran):</b> 
                    Hanbo Chen (2017). https://CRAN.R-project.org/package=VennDiagram
                    </br></br>
                    <b>GOglm:</b> 
                    https://github.com/gu-mi/GOglm
            </div>
            </span>
            <br/>
        </div>

    </body>
</html>

