<!DOCTYPE html>
<html>
<head>
    <title>GSEA Clusters Graph</title>
    <meta charset="UTF-8">
    <meta name="author" content="Ángeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>Project Data Visualization: GSEA Clusters Graph</b></span></div>
    <p>
        The Multidimensional GSEA Clusters Graph shows a network in which GO terms are clustered according to their functional
        similarity, i.e. the similarities in the biological processes (P), components (C) or broad molecular functions (F) that they 
        describe. The network is generated for the GO terms in the Multidimensional Enrichment Summary.
        In this network:
    </p>
    <ul>
        <li><b>Nodes</b> correspond to clustered GO terms. A label beside the node shows the GO term description.</li>
        <li><b>Edges</b> indicate the existence of semantic similarity between two GO terms.</li>
        <li><b>Node color</b> - in this network, each node will be colored in the form of a pie chart in order to compare the
        different levels of significance reported by MDGSA method for each GO term in each ranked list. Pie chart area 
        is defined by the relative -log10(over p-value) to measure the differential effect of each of the variables measured in 
        the ranked lists in the genes that are annotated under that GO term (i.e. a proxy to biological function).
        For instance, if DIU and DE genes are selected as input ranked lists, 
        different orange vs blue color proportions would point towards preferential regulation
        by Differential Isoform Usage or by Differential Expression mechanisms. However, different interpretations may be produced
        given that this analysis can be run for any kind of user-defined ranked list.</li>
        <li><b>Dynamic information</b> - use the mouse to scroll the graph to view full GO term names, GO IDs, number of genes
        significantly enriched at the top of the ranked lists, and p-values in each of the MDGSA result columns.</li>
        <li><b>Node size</b> indicates the number of genes annotated for that GO term that were detected as
        significantly highly ranked by MDGSA (i.e. by combining the two test lists).</li>
    </ul>
    <p>NOTE: Users may drag nodes and cluster labels to arrange the clusters and export high quality images of the clustering results
    (see left hand-side panel "Export" button).</p>
    <!-- DVReminders -->
    <!-- SubtabMenuBar:export -->
    <!-- AddHelp:fdsa,visual,menus,gui,tips -->
</body>
</html>

