<!DOCTYPE html>
<html>
<head>
    <title>Differential Feature Inclusion (DFI) Analysis parameters</title>
    <meta charset="UTF-8">
    <meta name="author" content="H. del Risco & Ángeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>Differential Feature Inclusion (DFI) Analysis parameters</b></span></div>
    <p>To evaluate Differential Feature Inclusion (DFI), you must provide the following information:
    </p>
    <ol class="dlgSection">
        <li>Annotation Feature(s) Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Annotation Feature(s) - </b>check to select the annotation feature(s) to include for DFI.
                Note that the tree contains one branch per annotation source and each branch can be expanded to show all the features it contains.
                You may select all features for a given annotation source or expand and select specific features only.
                You may include any number of features from multiple annotation sources, how ever, please note that including many categories may be significantly more time and memory-consuming.
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Check Using - </b>select which approach to use to compare features:
                <ul class="dlgFields">
                    <li>Feature presence - select to use the presence or absence of a feature for comparison. 
                        If feature A is found in one transcript and feature A is also found in another transcript, regardless of genomic position, the feature will not be considered for DFI.</li>
                    <li>Feature genomic position overlap - select to use the genomic position overlap of features for comparison.
                        Each instance of a feature, based on genomic location, is treated independently using this approach. 
			If feature A at location 1 is found in one transcript then feature A must also be found in the other transcript at an overlapping genomic position,
			the feature will not be considered for DFI.<br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Analysis Method Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Method - </b>select which analysis method to use for DFI:
                <ul class="dlgFields">
                    <li>DEXSeq - The package is focused on finding differential exon usage using RNA-seq exon counts between samples with different experimental designs. 
                        It provides functions that allows the user to make the necessary statistical tests 
                        based on a model that uses the negative binomial distribution to estimate the variance 
                        between biological replicates and generalized linear models for testing.</li>
		    <li>maSigPro - Differential isoform expression usage for RNA-seq data in single and multiple time-course experiments. After summing the expression of the feature-including and feature-excluding isoforms 
			for each feature, tappAS uses the Iso-maSigPro functionality in the maSigPro R package. maSigPro applies linear regression to model gene expression and selects significant genes via a two-step algorithm. 
			maSigPro returns lists of genes with statistically significant changes along time and across the different series.<br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Analysis Parameters Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Name - </b>enter a name for the analysis.
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Significance Level - </b>specify the significance level required for Differential Splicing (default: 0.05)
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Polynomial Degree - </b>(only time-course designs) degree for the maSigPro regression model. The maximum allowed degree is time_points-1.
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Filtering - </b>specify whether or not to filter minor isoforms prior to testing for Differential Isoform Usage. 
												Recommended to prevent spurious detection of DFI features due to small usage changes in minor isoforms. Two methods are available:
                <ul class="dlgFields">
                    <li>Filtering by proportion - <span style="font-weight: normal;">does not consider isoforms that represent a proportion of gene expression that is below the specified threshold (default: 0.2)</span><br/></li>
                    <li>Filtering by fold expression difference - <span style="font-weight: normal;">compares all isoforms to the major (i.e. the one with highest mean expression across conditions) and filters 
													those with a fold difference above the specified threshold (default: 2)</span><br/><br/></li>
                </ul>
	    </div>
        </li>
    </ol>
    <!-- AddHelp:dfi,edger,dexseq,gui,tips -->
    <br/><br/>
</body>
</html>

