<!DOCTYPE html>
<html>
<head>
    <title>Functional Diversity Analysis</title>
    <meta charset="UTF-8">
    <meta name="author" content="H. del Risco & Ãngeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>Functional Diversity Analysis</b></span></div>
    <p>The FDA Results table displays the analysis results. 
        The column sections, and columns, displayed will be determined by the annotation features selected for the FDA:</p>
    <ul>
        <li>Gene - gene identifier as provided in the annotation file.</li>
    </ul>
    <span style="padding-left:25px; font-weight:bold;">Transcript Annotation Features Section:</span>
    <ul>
        <li>Feature Category 1 - Result: Varying, Not Varying, or blank if gene does not contain annotation feature.</li>
        <li>...</li>
        <li>Feature Category n - Result: Varying, Not Varying, or blank if gene does not contain annotation feature.</li>
    </ul>
    <span style="padding-left:25px; font-weight:bold;">Protein Annotation Features Section:</span>
    <ul>
        <li>Feature Category 1 - Result: Varying, Not Varying, or blank if gene does not contain annotation feature.</li>
        <li>...</li>
        <li>Feature Category n - Result: Varying, Not Varying, or blank if gene does not contain annotation feature.</li>
    </ul>
    <!-- RowColumns -->
    <!-- HiddenColumns -->
    <ul>
        <li>Gene Description - gene description, if any, as provided in the annotation file.</li>
        <li>Annotation Feature Columns - use the 'Add annotation feature column...' menu selection.</li>
    </ul>

    <div><span class="contentTitle"><b>ID-Level Functional Diversity Analysis Results</b></span></div>
    <p>Analyzing Functional Diversity at the ID level generates a results table for each annotated element within the
	selected feature category (for instance, PFAM domains). One row per ID is displayed (for example, PF00009). 
	The columns contain the following:</p>
    <ul>
	<li>Feature ID: contains the ID of the feature as a composed name, integrated by the source (i.e. the database) from 
	which the annotation was retrieved, the annotation category name, and the feature name, separated by a bar. 
	Example: PFAM/DOMAIN/PF00009.</li>
    </ul>
    <ul>
        <li>Feature Description: a more detailed description of the annotated feature, if available.</li>
    </ul>
    <ul>
	<li>P-Value: significance of a Fisher's Exact test performed on varying/not varying status of the feature ID
	vs the annotation variation in the entire feature category.</li>
    </ul>
    <ul>
	<li>AdjP-Value: adjusted p-value using the Benjamini-Hochberg method.</li>
    </ul>
    <ul>
	<li>VARYING/NOT VARYING: number of genes for which the feature is detected as varying and not. The sum of these 
	two fields corresponds to the number of genes in the annotation that contain that specific feature in at least one isoform.</li>
    </ul>

    <!-- SubtabMenuBar:optlog,select,export,dataViz,rerun -->
    <!-- AddHelp:fda,tables,query,gui,tips -->
</body>
</html>

