<!DOCTYPE html>
<html>
<head>
    <title>Differential Isoform Usage analysis parameters</title>
    <meta charset="UTF-8">
    <meta name="author" content="H. del Risco & Ãngeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>Differential Isoform Usage analysis parameters</b></span></div>
    <p>To run Differential Isoform Usage (DIU) analysis, you must provide the following information:
    </p>
    <ol class="dlgSection">
        <li>Data Type Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Data Type - </b>select which gene data type to use for DIU:
                <ul class="dlgFields">
                    <li>CDS - <span style="font-weight: normal;">use gene CDS expression data (obtained after collapsing transcript isoform expression by coding region similarity)</span><br/></li>
                    <li>Transcripts - <span style="font-weight: normal;">use gene transcript expression data</span><br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Analysis Method Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Method - </b>select which analysis method to use for DIU:
                <ul class="dlgFields">
                    <li>DEXSeq - The package is focused on finding differential exon usage using RNA-seq exon counts between samples with different experimental designs. 
                        It provides functions that allows the user to make the necessary statistical tests 
                        based on a model that uses the negative binomial distribution to estimate the variance 
                        between biological replicates and generalized linear models for testing.<br/><br/></li>
	            <li>maSigPro - Differential isoform expression usage for RNA-seq data in single and multiple time-course experiments. Uses the Iso-maSigPro functionality
			in the maSigPro R package. maSigPro applies linear regression to model gene expression and selects significant genes via a two-step algorithm. 
			First, responsive genes are identified by fitting a generic regression model with time as quantitative variable and series as dummy variables. 
			Second, step-wise regression is applied on selected genes to adjust models and identify gene-specific variation patterns. 
			maSigPro returns lists of genes with statistically significant changes along time and across the different series.<br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Analysis Parameters Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Significance Level - </b>specify the significance level required for Differential Isoform Usage (default: 0.05)
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Polynomial Degree - </b>(only time-course designs) degree for the maSigPro regression model. The maximum allowed degree is time_points-1.
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Filtering - </b>specify whether or not to filter minor isoforms prior to testing for Differential Isoform Usage. 
												Recommended to prevent spurious detection of DIU genes due to small usage changes in minor isoforms. Two methods are available:
                <ul class="dlgFields">
                    <li>Filtering by proportion - <span style="font-weight: normal;">does not consider isoforms that represent a proportion of gene expression that is below the specified threshold (default: 0.2)</span><br/></li>
                    <li>Filtering by fold expression difference - <span style="font-weight: normal;">compares all isoforms to the major (i.e. the one with highest mean expression across conditions) and filters 
													those with a fold difference above the specified threshold (default: 2)</span><br/><br/></li>
                </ul>
	    </div>

        </li>
    </ol>
    <!-- AddHelp:dsa,edger,dexseq,gui,tips -->
    <br/><br/>
</body>
</html>

