<!DOCTYPE html>
<html>
<head>
    <title>Differential Expression Analysis Parameters</title>
    <meta charset="UTF-8">
    <meta name="author" content="H. del Risco & Ángeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>Differential Expression Analysis Parameters</b></span></div>
    <p>To run Differential Expression Analysis (DEA), you must provide the following information:
    </p>
    <ol class="dlgSection">
        <li>Data Type Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Data Type - </b>select which data type to use for DEA:
                <ul  class="dlgFields">
                    <li>Genes - <span style="font-weight: normal;">use gene expression data (obtained after collapsing transcript isoform expression for each gene)</span><br/></li>
                    <li>Transcripts - <span style="font-weight: normal;">use transcript expression data</span><br/></li>
                    <li>CDS - <span style="font-weight: normal;">use CDS expression data (obtained after collapsing transcript isoform expression by coding region similarity)</span><br/></br></li>
                </ul>
            </div>
        </li>
        <li>Analysis Method Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Method - </b>select which analysis method to use for DEA:
                <ul class="dlgFields">
                    <li>edgeR - Differential expression analysis of RNA-seq expression profiles with biological replication. 
                        Implements a range of statistical methodology based on the negative binomial distributions, 
                        including empirical Bayes estimation, exact tests, generalized linear models and quasi-likelihood tests.</li>
                    <li>NOISeq - Differential expression between two experimental conditions with no parametric assumptions.<br/><br/></li>
                    <li>maSigPro - Differential expression for RNA-seq data in single and multiple time-course experiments. 
			Applies linear regression to model gene expression and selects significant genes via a two-step algorithm. 
			First, responsive genes are identified by fitting a generic regression model with time as quantitative variable and series as dummy variables. 
			Second, step-wise regression is applied on selected genes to adjust models and identify gene-specific variation patterns. 
			maSigPro returns lists of genes with statistically significant changes along time and across the different series.<br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Analysis Parameters Section (two-group comparisons)
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Replicates - </b>select the type of replicates provided in the expression matrix, if any (NOISeq only):
                <ul class="dlgFields">
                    <li>Biological - select for biologically distinct samples</li>
                    <li>Technical - select for same samples measured using different techniques</li>
                </ul>
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Significance Level - </b>specify the significance level required for Differential Expression (default: 0.05)
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Fold Change - </b>specify the Fold Change required for Differential Expression (default: 2, NOISeq only)
            <br/><br/></div>
        <li>Analysis Parameters Section (single and multiple time course designs)
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Polynomial Degree - </b>degree for the regression model. The maximum allowed degree is time_points-1.
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Significance Level - </b>specify the significance level required for Differential Expression (default: 0.05)
            </div>
	    <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>R^2 cutoff - </b>required level of the goodness of fit of the regression model. This parameter is between 0 and 1. 
											Higher values indicate well fitted models. MaSigPro documentation recommends values between [0.4,0.8].
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Max K Clusters - </b>To show the trajectories in plots maSigPro visualization applies clustering methods to group 
											genes with similar trends and summarize the graphical display. This parameter determines the maximum number of clusters to classify genes into.
											Enable the "Use mclust for optimal number" option to use a method to dynamically determine the number of clusters.
            </div>
        </li>
    </ol>
    <!-- AddHelp:dea,edger,noiseq,gui,tips -->
    <br/><br/>
</body>
</html>

