<!DOCTYPE html>
<html>
<head>
    <title>Functional Enrichment Analysis parameters</title>
    <meta charset="UTF-8">
    <meta name="author" content="H. del Risco & Ángeles Arzalluz-Luque">
    <link href="css/web.css" rel="stylesheet" type="text/css">
    <script src="scripts/scripts.js"></script>
</head>

<body style="padding:5px; margin:0px; border:0px; border-spacing:0px;">
    <div><span class="contentTitle"><b>Functional Enrichment Analysis parameters</b></span></div>
    <p>To run Functional Enrichment Analysis (FEA), you must provide the following information:
    </p>
    <ol class="dlgSection">
        <li>Analysis Information Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Name - </b>specify unique name for analysis
            </div><br/>
        </li>
        <li>Data Type Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Data Type - </b>select which data type to use for FEA:
                <ul  class="dlgFields">
                    <li>Genes - <span style="font-weight: normal;">use gene expression data (obtained after collapsing expression of isoforms from the same gene)</span><br/></li> 
                    <li>Transcripts - <span style="font-weight: normal;">use transcript expression data</span><br/></li>
		    <li>CDS - <span style="font-weight: normal;">use CDS expression data (obtained after collapsing expression of isoforms with the same CDS)</span><br/><br/></li>
                </ul>
            </div>
        </li>
        <li>Item Lists Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Test List - List of items (gene/transcript/CDS) whose enrichment will be evaluated.
												  Select from one of the options provided (will vary based on data type).</b>
                <div style="padding-top:7px; padding-left:10px;">
                    For genes:
                    <ul class="dlgFields">
                        <li>Differential Isoform Usage (Transcripts) - DIU genes based on transcripts</li>
                        <li>Not Differential Isoform Usage (Transcripts) - not DIU genes based on transcripts</li>
                        <li>Differential Isoform Usage (Proteins) - DIU genes based on proteins</li>
                        <li>Not Differential Isoform Usaged (Proteins) - not DIU genes based on proteins</li>
                        <li>Differentially Expressed - differentially expressed genes</li>
                        <li>Not Differentially Expressed - not differentially expressed genes</li>
                        <li>Use list from file - specify a list to use in file selection control below</li> 
                    </ul>
                    For transcripts:
                    <ul class="dlgFields">
                        <li>Differentially Expressed - differentially expressed transcripts</li>
                        <li>Not Differentially Expressed - not differentially expressed transcripts</li>
                        <li>Use list from file - specify a list to use in file selection control below</li> 
                    </ul>
                    For CDS:
                    <ul class="dlgFields">
                        <li>Differentially Expressed - differentially expressed CDS</li>
                        <li>Not Differentially Expressed - not differentially expressed CDS</li>
                        <li>Use list from file - specify a list to use in file selection control below</li> 
                    </ul>
                    <div style="padding-top:5px;"><span class="notes">Note that you must have previously generated the data for some of the options, e.g. DIU and DEA results.</span></div>
                </div>
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Background List - List of items (gene/transcript/CDS) against which enrichment will be tested.
												  Select from one of the options provided (will vary based on data type).</b>
                <div style="padding-top:7px; padding-left:10px;">
                    For genes:
                    <ul class="dlgFields">
                        <li>All Genes - all genes in project</li>
                        <li>All multiple isoforms genes - all genes in projects with multiple isoforms</li>
                        <li>Use list from file - specify a list to use in file selection control below</li> 
                    </ul>
                    For transcripts:
                    <ul class="dlgFields">
                        <li>All Transcripts - all transcripts in project</li>
                        <li>Use list from file - specify a list to use in file selection control below</li> 
                    </ul>
                    For CDS:
                    <ul class="dlgFields">
                        <li>All CDS - all CDS in project</li>
                        <li>Use list from file - specify a list to use in file selection control below</li> 
                    </ul>
                </div>
            </div><br/>
        </li>
        <li>Annotation Feature(s) Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Annotation Feature(s) - </b>select the annotation feature(s) to include for FEA.
                Note that the tree contains one branch per annotation source and each branch can be expanded to show all the features it contains.
                You may select all features for a given annotation source or expand and select specific features only.
                You may include features from a several annotation sources. 
            </div><br/>
        </li>
        <li>Analysis Parameters Section
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Statistical Test - </b>select which statistical test to use:
                <ul class="dlgFields">
                    <li>Wallenius - approximates the true distribution of numbers of members of a category amongst DE
			genes by the Wallenius non-central hypergeometric distribution.</li>
                    <li>Sampling - uses random sampling to approximate the true distribution and uses it to calculate the p-
			values for over (and under) representation of categories. Must specify the number of random samples to be calculated if so.</li>
                    <li>Hypergeometric - Hypergeometric assumes there is no bias in power to detect differential expression at all and
			calculates the p-values using a standard hypergeometric distribution. Useful if you wish to test the
			effect of selection bias on your results. CAUTION: Hypergeometric should NEVER be used for producing results for biological interpretation. </li>
                </ul>
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Use w/o Category - </b>select whether to use items not containing specific features or not:
                <ul class="dlgFields">
                    <li>Yes - include items even if they do not contain a specific feature being tested. These genes will count
			towards the total number of genes outside the feature being tested.</li>
                    <li>No - only include items containing specific features tested. These genes will be ignored in the calculation of p-values
			(default).</li>
                </ul>
                <div style="padding-top:5px;"><span class="notes" style="padding-top:12px;">Note that only items that have at least one of the features selected will be included to begin with.
                        However, it is possible for an item to have some of the selected features but not all.</span></div>
            </div>
            <div style="padding-top:7px; padding-left:10px; font-weight: normal; color:black;"><b>Significance Level - </b>specify the significance level required for Differential Expression (default: 0.05)
            </div>
        </li>
    </ol>
    <!-- AddHelp:fea,goseq,gui,tips -->
    <br/><br/>
</body>
</html>

