<!DOCTYPE html>
<html>
<head>
    <title>Functional Enrichment Analysis (FEA)</title>
	<meta charset="UTF-8">
	<meta name="author" content="H. del Risco">
	<link rel="stylesheet" type="text/css" href="css/web.css">
	<script src="scripts/scripts.js"></script>
	<script src="scripts/slides.js"></script>
	<script src="scripts/maps.js"></script>
	<script src="scripts/ebook.js"></script>
	<script src="scripts/overview.js"></script>
</head>

<body style="padding-top:0px; margin-left:0px; margin-right:0px; margin-top:0px; height:100%; min-width:620px; width:100%; overflow:auto;" onresize="onResize()" onload="init(false, 'pageEA', 'Enrichment Analysis')" onhashchange="onHashChanged()">


  <section id="mainSection" style="position:fixed; left:0px; top:0px; right:0; bottom:0px; display:block; overflow:auto;">
	<section id="contentSection" style="float:left; margin:auto; width:400px; min-width:725px; display:inline-block;">
<div id="Top" style="height:12px;"></div>
<table style="width:100%; vertical-align:middle;">
  <tr>
    <td style="text-align:left;"><span id="contentTitle" class="contentTitle">Introduction</span></td>
    <td style="width:25px;">
        <div id="pagePrev" class="contentButtonOff" onclick="ebook.pagePrev()">&lt;</div>
    </td>
    <td style="width:25px;">
        <div id="pageNext" class="contentControl contentButton" onclick="ebook.pageNext()">&gt;</div>
    </td>
    <td id="tdContentControl" style="width:140px; text-align:right;">
        <div id="contentControl" class="contentControl" style="padding-left:5px; padding-right:5px;" onclick="toggleContents()">Hide Contents &gt;&gt;</div>
    </td>
  </tr>
</table>

		<section id="pageEA" class="contentPage">
<p>Several enrichment analysis algorithms are implemented in tappAS, and can be run on all available functional annotation databases, performing both traditional GO term, pathway, etc. enrichment analysis
as well as enrichments considering any other functions.
As in the rest of analysis, users may specify what data type to use - genes, transcripts or CDS - for the analysis. In addition, both the test and background lists to use can be user-defined, 
however, available application lists are provided (i.e. DE genes, DIU genes...). 
Users also need to specify annotation categories for testing as well as some analysis package required parameters (see Help button on the bottom of the dialogue for instructions).
For all enrichment analysis, several categories can be simultaneously analyzed.</p>

<br/><br/>
<div><span id="FEA" class="contentTitle">Functional Enrichment Analysis (FEA)</span></div>
<p>Functional Enrichment Analysis measures the significance of the overrepresentation of a characteristic (GO term annotation, fucntional feature, etc.) in a test list of elements (genes, transcripts or CDS), 
in comparison with the proportion in which this characteristic is present in a background list. TappAS currently uses the GOSeq R package as the underlying method for Functional Enrichment.
You may view the documentation and installation instructions at: <a href="https://www.bioconductor.org/packages/release/bioc/html/goseq.html" target="_blank" style="padding-left:30px;">
https://www.bioconductor.org/packages/release/bioc/html/goseq.html</a>
	
</p>
<br/><br/>

<br/><br/>
<div style="text-align:center; margin:auto; padding-left:20px; padding-right:20px;">
    <img src="images/FEAdialog.png" style="max-width:100%; max-height:100%;" alt="FEA Parameters"/><br/><br/>
    <div>Functional Enrichment Analysis Parameters</div>
</div>
<br/><br/>
When preparing to run the analysis, all FEA parameters are described in the Help page, which can be accessed via the Help button located on the bottom of the dialog window.

<br/><br/>
<div>
	<b>FEA Results</b><br/>
	<p>After running the analysis, the FEA results table will be displayed. This table is contained in the project data tab, located in the top tab panel (see image below). 
	Each table row displays the results for a tested feature: Significant - Yes/No, overrepresentation and underrepresentation p-values, and multiple-testing adjusted P-Values.
	The number of test genes and total genes containing the selected feature are also shown. Data drill down is available for these lists by clicking on the right mouse button.</p>
</div><br/>
<div style="text-align:center; margin:auto; padding-left:20px; padding-right:20px;">
    <img src="images/FEAresults.png" style="max-width:100%; max-height:100%;" alt="FEA Results"/><br/><br/>
    <div>Functional Enrichment Analysis Results</div>
</div>
<br/><br/>
<div><span id="FEAClusters" class="contentTitle">Results Summary and Enriched Features Chart</span></div>
<p>After performing a Functional Enrichment, users may click the visualization button on the left hand side panel to display the Results Summary plots, or to produce an Enriched Features
Chart summarizing the significance level and the number of genes in the test list for all significant features (bottom panel). Note that this chart can be explored by using the Help button and the visualization options
in the left hand side bar.</p>
<br/><br/>

<div><span id="FEAClusters" class="contentTitle">Enriched Features Cluster Analysis</span></div>
<p>When performing a Functional Enrichment of GO terms, the option to run Cluster Analysis on the FEA results is provided in the left hand side menu by clicking on the clusters icon.
An example of this cluster analysis results can be seen below: the table on the left displays the generated clusters, while the table on the right shows the IDs and descriptions of the GO term 
nodes for the selected cluster(s). You may select multiple clusters to see their combined nodes. A results summary of the clusters can be displayed on the bottom panel by clicking on the visualization
icon on the left hand side menu.</p>
<div style="text-align:center; margin:auto; padding-left:20px; padding-right:20px;">
    <img src="images/FEAclusters.png" style="max-width:100%; max-height:100%;" alt="FEA Results"/><br/><br/>
    <div>FEA Results - Enriched Features Cluster Analysis</div>
</div>
<br/><br/><br/>

<div><span id="GSEA" class="contentTitle">Gene Set Enrichment Analysis (GSEA)</span></div>
<p>Gene Set Enrichment Analysis tests for the significance of overrepresentation of certain characteristic (for instance, GO terms, or any functional feature in tappAS) by incorporating 
a ranking metric (for instance, significance) for the gene/transcript/CDS identifiers in the relevant gene list. As previously stated, available application ranked lists are provided, 
but users may use any custom ranked list file. Several functional categories can be simultaneously run for both implemented methods.
</p>

<br/><br/>
<div style="text-align:center; margin:auto; padding-left:20px; padding-right:20px;">
    <img src="images/GSEAdialog.png" style="max-width:100%; max-height:100%;" alt="GSEA Parameters"/><br/><br/>
    <div>Differential Feature Inclusion Analysis Parameters</div>
</div>
<br/><br/>
When preparing to run the analysis, all GSEA parameters are described in the Help page, which can be accessed via the Help button located on the bottom left of the dialog window.
There are two GSEA methods implemented in tappAS:
<br/><br/>
<div>
	<b>GOglm</b><br/>
	<p>GOglm provides a length bias-corrected approach for GSEA.</p>
	You may view the documentation and installation instructions at:<br/>
	<div style="padding-top:7px;">
	<a href="https://github.com/gu-mi/GOglm" target="_blank" style="padding-left:30px;">https://github.com/gu-mi/GOglm</a>
	</div>
</div>
<br/>
<div>
	<b>MDGSA</b><br/>
	<p>MDGSA provides functions to perform GSEA in several genomic dimensions. Note that the MDGSA can be run for a single ranked list (unidimensional), or combining two ranked lists 
	(multidimensional). Importantly, multidimensional MDGSA in tappAS is ONLY enabled for GO term enrichment.</p>
	You may view the documentation and installation instructions at:<br/>
	<div style="padding-top:7px;">
	<a href="https://github.com/gu-mi/GOglm" target="_blank" style="padding-left:30px;">https://github.com/gu-mi/GOglm</a>
	</div>
</div>
<br/>
<br/><br/>
<div>
	<b>GSEA Results</b><br/>
	<p>The GSEA results table is contained in the project data tab, located in the top tab panel (see image below), and will be displayed immediately after running the analysis. 
	Each table row displays the results (i.e. Significant - Yes/No) for a given annotation feature, as well as overrepresentation and underrepresentation test p-values and adjusted P-Values.
	As in the rest of analysis, more columns may be added using the (+) button on the right hand side of the table.</p>

<div style="text-align:center; margin:auto; padding-left:20px; padding-right:20px;">
    <img src="images/GSEAresults.png" style="max-width:100%; max-height:100%;" alt="GSEA Results"/><br/><br/>
    <div>Gene Set Enrichment Analysis Results</div>
</div>

	<p>For multidimensional MDGSA, the visualization button will display the "View Multidimensional Enrichment Summary", which will output a list filtered by a user-defined list or a specified significance
	threshold. This table can be used to produce a multidimensional MDGSA Result Summary visualization by clicking on the visualization button on the left hand side menu. This new graph will appear
	in the bottom panel. Use its Help button to get more information on this graph.</p>
</div><br/>

<br/><br/>

		</section>
	</section>
<aside id="contentItems" class="contentMenuApp">
    <div style="height:19px;"></div>
    <span style="color:slategray; font-size:medium; font-weight:bold;">Contents<br/></span>
    <div style="height:3px;"></div>
    <div id="pageIntro_Top" class="contentItem"><a href="pageIntro.html#Top" class="notSelected">Introduction</a></div>
    <div id="pageIntro_subitems" class="contentSubItems">
            <div id="pageIntro_Reqs" class="contentItem" style="padding-left:10px;"><a href="pageIntro.html#Reqs" class="notSelected">Application Requirements</a></div>
    </div>
    <div id="pageProjects_Top" class="contentItem"><a href="pageProjects.html#Top" class="notSelected">Projects</a></div>
    <div id="pageProjects_subitems" class="contentSubItems">
            <div id="pageProjects_InputData" class="contentItem" style="padding-left:10px;"><a href="pageProjects.html#InputData" class="notSelected">Input Data and Filtering</a></div>
            <div id="pageProjects_Normalization" class="contentItem" style="padding-left:10px;"><a href="pageProjects.html#Normalization" class="notSelected">Expression Data Normalization</a></div>
            <div id="pageProjects_DesignFileFormat" class="contentItem" style="padding-left:10px;"><a href="pageProjects.html#DesignFileFormat" class="notSelected">Experiment Design File Format</a></div>
            <div id="pageProjects_ExpMatrixFileFormat" class="contentItem" style="padding-left:10px;"><a href="pageProjects.html#ExpMatrixFileFormat" class="notSelected">Expression Matrix File Format</a></div>
            <div id="pageProjects_AnnotationFileFormat" class="contentItem" style="padding-left:10px;"><a href="pageProjects.html#AnnotationFileFormat" class="notSelected">Annotation File Format</a></div>
    </div>
    <div id="pageAppXface_Top" class="contentItem"><a href="pageAppXface.html#Top" class="notSelected">Application Interface</a></div>
    <div id="pageAppXface_subitems" class="contentSubItems">
            <div id="pageAppXface_Layout" class="contentItem" style="padding-left:10px;"><a href="pageAppXface.html#Layout" class="notSelected">GUI Layout</a></div>
            <div id="pageAppXface_Menus" class="contentItem" style="padding-left:10px;"><a href="pageAppXface.html#Menus" class="notSelected">Context-Sensitive Menus</a></div>
            <div id="pageAppXface_Tabs" class="contentItem" style="padding-left:10px;"><a href="pageAppXface.html#Tabs" class="notSelected">Tab Panels, Tabs, and Subtabs</a></div>
            <div id="pageAppXface_SubtabMenuBar" class="contentItem" style="padding-left:10px;"><a href="pageAppXface.html#SubtabMenuBar" class="notSelected">Subtabs Menu Bar</a></div>
            <div id="pageAppXface_Tables" class="contentItem" style="padding-left:10px;"><a href="pageAppXface.html#Tables" class="notSelected">Tables</a></div>
            <div id="pageAppXface_Visual" class="contentItem" style="padding-left:10px;"><a href="pageAppXface.html#Visual" class="notSelected">Visual Display Controls</a></div>
    </div>
    <div id="pageAppTips_Top" class="contentItem"><a href="pageAppTips.html#Top" class="notSelected">Application Tips</a></div>
    <div id="pageDataViz_Top" class="contentItem"><a href="pageDataViz.html#Top" class="notSelected">Data Visualization</a></div>
    <div id="pageDataViz_subitems" class="contentSubItems">
            <div id="pageDataViz_DVAccess" class="contentItem" style="padding-left:10px;"><a href="pageDataViz.html#DVAccess" class="notSelected">Accessing Data Visualizations</a></div>
            <div id="pageDataViz_GeneDV" class="contentItem" style="padding-left:10px;"><a href="pageDataViz.html#GeneDV" class="notSelected">Gene Data Visualization</a></div>
    </div>
    <div id="pageDrillDown_Top" class="contentItem"><a href="pageDrillDown.html#Top" class="notSelected">Data Drill Down</a></div>
    <div id="pageExport_Top" class="contentItem"><a href="pageExport.html#Top" class="notSelected">Export Data and Images</a></div>
    <div id="pageQuery_Top" class="contentItem"><a href="pageQuery.html#Top" class="notSelected">Ad Hoc Query</a></div>
    <div id="pageQuery_subitems" class="contentSubItems">
            <div id="pageQuery_Search" class="contentItem" style="padding-left:10px;"><a href="pageQuery.html#Search" class="notSelected">Simple Search</a></div>
            <div id="pageQuery_Query" class="contentItem" style="padding-left:10px;"><a href="pageQuery.html#Query" class="notSelected">Row Selection Query</a></div>
    </div>
    <div id="pageAFA_Top" class="contentItem"><a href="pageAFA.html#Top" class="notSelected">Functional Diversity Analysis</a></div>
    <div id="pageAFA_subitems" class="contentSubItems">
            <div id="pageAFA_Gene" class="contentItem" style="padding-left:10px;"><a href="pageAFA.html#Gene" class="notSelected">Gene-Level FDA</a></div>
            <div id="pageAFA_ID" class="contentItem" style="padding-left:10px;"><a href="pageAFA.html#ID" class="notSelected">ID-Level FDA</a></div>
            <div id="pageAFA_Combined" class="contentItem" style="padding-left:10px;"><a href="pageAFA.html#Combined" class="notSelected">ID-Level FDA Combined Results</a></div>
    </div>
    <div id="pageDEA_Top" class="contentItem"><a href="pageDEA.html#Top" class="notSelected">Differential Expression Analysis</a></div>
    <div id="pageDIU_Top" class="contentItem"><a href="pageDIU.html#Top" class="notSelected">Differential Isoform Usage Analysis</a></div>
    
    <div id="pageEA_Top" class="contentItem"><a href="pageEA.html#Top" class="notSelected">Enrichment Analysis</a></div>
    <div id="pageEA_subitems" class="contentSubItems">
            <div id="pageEA_FEA" class="contentItem" style="padding-left:10px;"><a href="pageEA.html#FEA" class="notSelected">Functional Enrichment Analysis</a></div>
            <div id="pageEA_FEAClusters" class="contentItem" style="padding-left:10px;"><a href="pageEA.html#FEAClusters" class="notSelected">Enriched Features Clusters</a></div>
            <div id="pageEA_GSEA" class="contentItem" style="padding-left:10px;"><a href="pageEA.html#GSEA" class="notSelected">Gene Set Enrichment Analysis</a></div>
    </div>
</aside>
<div style="clear:both;"></div>

  </section>

</body>
</html>